happycache.c

```diff
@@ -110,86 +112,82 @@ int prepare_file(char * line, fd_info * fds) {
 		return fd;
 	}
 
-	if(posix_madvise(fdi->base_addr, file_stat.st_size, POSIX_MADV_RANDOM) != 0) {
-		fprintf(stderr, "Could not posix_madvise %s: ", line);
-		perror(NULL);
-		return fd;
-	}
-
 	//Only update the struct's num_pages once we're confident everything's
 	//tip-top
 	fdi->num_pages = num_pages;
 
+	uint32_t chunk_pages = num_pages > CHUNK_SIZE ? CHUNK_SIZE : num_pages;
+	fdi->mincore = malloc(chunk_pages);
+	mincore(fdi->base_addr, chunk_pages * page_size, fdi->mincore);
 	return fd;
 }
 
-void enqueue_load_page(
-	fd_info * fds,
+void load_pages(
+	fd_info * fdi,
 	int fd,
-	int64_t page,
-	sll * free_list,
-	sll * work_list
+	uint64_t start,
+	uint64_t count
 ) {
 	if(fd == -1) {
 		return;
 	}
 
-	char * file_mmap = fds[fd].base_addr;
-	uint64_t num_pages = fds[fd].num_pages;
+	uint64_t num_pages = fdi->num_pages;
 
-	if(page > num_pages) {
-		return;
+	if(num_pages - start < count) {
+		count = num_pages - start;
 	}
-	read_work * rw = container_of(
-		list_pop_head(free_list),
-		read_work,
-		list
-	);
-	finished_file_op(&fds[rw->fd], rw->fd);
 
-	fds[fd].refcount++;
-	rw->fd = fd;
-	rw->addr = (void *)&file_mmap[page << page_shift];
-	rw->base_addr = file_mmap;
-	list_push_tail(work_list, &rw->list);
-}
+	while(count > 0) {
+		uint64_t mincore_offset = start - fdi->mincore_start;
+		if(mincore_offset > CHUNK_SIZE) {
+			fdi->mincore_start = start;
+			mincore_offset = 0;
+			uint64_t remaining = num_pages - start;
+			uint64_t chunk_pages = remaining > CHUNK_SIZE ? CHUNK_SIZE : remaining;
+
+			mincore(
+				fdi->base_addr + fdi->mincore_start * page_size,
+				chunk_pages * page_size,
+				fdi->mincore
+			);
+		}
 
-int load_from_map(gzFile map, int num_threads) {
-	sll work_list;
-	sll free_list;
+		if((fdi->mincore[mincore_offset] & 0x01) == 0) {
+			//Page is not already in RAM.
+			break;
```

Comment:
Does this mean that if there are remaining pages after this we won't load them?

Reply:
It's kinda the opposite. The purpose of this loop is to skip over (i.e. not submit readahead requests for) pages which are already in RAM. If we hit the `break` statement, it is because we found at least one page already in RAM. Once we've broken out of the loop, all `count` pages in the range will be loaded via posix_fadvise, even if they were already there.

I suppose it could be more efficient, but it's nbd.